### PR TITLE
Allow implicit casts in switch labels; improve error message

### PR DESCRIPTION
**Pull request by mbudiu-vmw**
Signed-off-by: Mihai Budiu <mbudiu@vmware.com>

### CODE REVIEW

**Comment by mbudiu-vmw** [comment]
Fixes #3623

File: frontends/p4/typeChecking/typeChecker.cpp [hunk 1]
```diff
@@ -3836,6 +3836,8 @@ const IR::Node* TypeInference::postorder(IR::SwitchStatement* stat) {
             auto lt = getType(c->label);
             if (lt == nullptr)
                 continue;
+            if (auto se = lt->to<IR::Type_SerEnum>())
+                lt = se->type;
```
Comment (apinski-cavium) [inline frontends/p4/typeChecking/typeChecker.cpp:3840]: I am not 100% sure with this change but will the following still work:
```
enum bit<4> e
{
  a = 1
}
control c(in e v)
{
  apply{
    switch(v) {
      e.a:
      default: {}
    }
  }
}
```
and will this still be rejected:
```
enum bit<4> e
{
  a = 1
}
control c(in e v)
{
  apply{
    switch(v) {
      1:
      default: {}
    }
  }
}
```

Comment (mbudiu-vmw) [reply frontends/p4/typeChecking/typeChecker.cpp]: I added theses test cases.

File: testdata/p4_16_errors_outputs/issue3623-2.p4-stderr [hunk 2]
```diff
@@ -0,0 +1,6 @@
+issue3623-2.p4(8): [--Werror=type-error] error: 'v' with type 'e' cannot be compared to '4w1' with type 'bit<4>'
```
Comment (fruffy) [inline testdata/p4_16_errors_outputs/issue3623-2.p4-stderr:1]: This error message are a little vague. Any way to make it more clear that this concerns a (ser)enum?

**Review by fruffy — APPROVED** [review]

File: testdata/p4_16_errors_outputs/issue3623-2.p4-stderr (hunk 2 above)
Comment (mbudiu-vmw) [reply testdata/p4_16_errors_outputs/issue3623-2.p4-stderr:1]: Removing 'toString' will show better error messages. Just pushed this change.

**Pull request merged**